File: src/main/java/com/github/syl20lego/excel2json/Executor.java
package com.github.syl20lego.excel2json;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import com.fasterxml.jackson.databind.SerializationFeature;
import org.apache.poi.openxml4j.exceptions.InvalidFormatException;
import org.apache.poi.ss.usermodel.*;

import java.awt.*;
import java.io.*;
import java.util.*;
import java.util.List;

public class Executor {

    public static void main(String[] args) throws InvalidFormatException {
        try {
            FileDialog dialog = new FileDialog((Frame) null, "Select File to Open");
            dialog.setMode(FileDialog.LOAD);
            dialog.setFilenameFilter((dir, name) -> name.endsWith(".xlsx") || name.endsWith(".xls"));
            dialog.setVisible(true);
            String file = dialog.getFile();
            String directory = dialog.getDirectory();
            if (file != null) {
                String jsonFile = file.substring(0, file.lastIndexOf(".")) + ".json";

                Workbook workbook = WorkbookFactory.create(new File(directory + file), null, true);

                Map<String, List> sheets = new LinkedHashMap<>();
                workbook.forEach(sheet -> {
                    List<Map<String, Object>> list = new ArrayList<>();
                    sheets.put(sheet.getSheetName(), list);
                    List<String> header = new ArrayList<>();
                    sheet.forEach(row -> {
                        Map<String, Object> outputRow = new LinkedHashMap<>();
                        boolean headerRow = header.isEmpty();
                        row.forEach(cell -> {
                            if (headerRow) {
                                header.add(cell.toString());
                            } else {
                                if (cell.toString().length() > 0) {
                                    outputRow.put(header.get(cell.getColumnIndex() - 1), cellValue(cell));
                                }
                            }
                        });
                        if (!outputRow.isEmpty()) {
                            list.add(outputRow);
                        }
                    });

                });


                ObjectMapper mapper = new ObjectMapper();
                mapper.enable(SerializationFeature.INDENT_OUTPUT);
                mapper.getFactory().configure(JsonGenerator.Feature.ESCAPE_NON_ASCII, true);
                ObjectWriter writer = mapper.writer(new DefaultPrettyPrinter());

                BufferedWriter fileWriter = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(directory + jsonFile), "Cp1252"));
                writer.writeValue(fileWriter, sheets);

                // Closing the workbook
                workbook.close();
            }

            System.out.println("Done");
        } catch (IOException e) {
            e.printStackTrace();
            System.out.println("Expecting file xlsx");
        }
    }

    private static String cellValue(Cell cell) {
        DataFormatter dataFormatter = new DataFormatter();

        switch (cell.getCellTypeEnum()) {
            case STRING:
                return cell.getRichStringCellValue().getString();
            case NUMERIC:
                if (DateUtil.isCellDateFormatted(cell)) {
                    return dataFormatter.formatCellValue(cell);
                } else {
                    if ((cell.getNumericCellValue() % 1) == 0) {
                        return Integer.valueOf(Double.valueOf(cell.getNumericCellValue()).intValue()).toString();
                    } else {
                        return Double.valueOf(cell.getNumericCellValue()).toString();
                    }
                }
            case FORMULA:
                return cell.getCellFormula();
            case BLANK:
                return "";
            default:
                return "";
        }
    }
}

File: src/main/java/com/github/syl20lego/excel2json/Executor.kt
package com.github.syl20lego.excel2json

import com.fasterxml.jackson.core.util.DefaultPrettyPrinter
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import org.apache.poi.ss.usermodel.*

import java.awt.*
import java.io.*
import java.util.*

data class FileInfo(val directory: String?, val file: String?)

private fun open(): FileInfo {
    val dialog = FileDialog(null as Frame?, "Select File to Open")
    dialog.mode = FileDialog.LOAD
    dialog.setFilenameFilter { dir, name -> name.endsWith(".xlsx") || name.endsWith(".xls") }
    dialog.isVisible = true
    return FileInfo(dialog.directory, dialog.file)
}

private fun cellValue(cell: Cell?): String {

    cell ?: return ""
    when (cell.cellTypeEnum) {
        CellType.STRING -> return cell.richStringCellValue.string
        CellType.FORMULA -> return cell.cellFormula
        CellType.BLANK -> return ""
        CellType.NUMERIC -> return if (DateUtil.isCellDateFormatted(cell)) {
            val dataFormatter = DataFormatter()
            dataFormatter.formatCellValue(cell)
        } else {
            if (cell.numericCellValue % 1 == 0.0) {
                Integer.valueOf(java.lang.Double.valueOf(cell.numericCellValue).toInt())!!.toString()
            } else {
                java.lang.Double.valueOf(cell.numericCellValue).toString()
            }
        }
        else -> return ""
    }
}

private fun writeJson(directory: String?, file: String, sheets: LinkedHashMap<String, List<*>>) {
    val jsonFile = file.substring(0, file.lastIndexOf(".")) + ".json"
    val mapper = ObjectMapper()
    mapper.enable(SerializationFeature.INDENT_OUTPUT)
    val writer = mapper.writer(DefaultPrettyPrinter())
    val fileWriter = BufferedWriter(OutputStreamWriter(FileOutputStream(directory + jsonFile), "UTF-8"))
    writer.writeValue(fileWriter, sheets)
}

fun isNotEmptyRow(row: Row): Boolean {
    for (cellNum in row.firstCellNum until row.lastCellNum) {
        val cell = row.getCell(cellNum)
        if (cell != null &&  cell.cellTypeEnum != CellType.BLANK && cell.toString().isNotEmpty()) {
            return true
        }
    }
    return false
}

object Executor {

    @JvmStatic
    fun main(args: Array<String>) {
        val (directory, file) = open()
        file ?: return

        val sheets = LinkedHashMap<String, List<*>>()
        WorkbookFactory.create(File(directory + file), null, true)
                .use {
                    it.forEach { sheet ->
                        val list = ArrayList<Map<String, Any>>()
                        sheets[sheet.sheetName] = list
                        val header = ArrayList<String>()
                        sheet.forEach { row ->
                            val outputRow = LinkedHashMap<String, Any>()
                            if (header.isEmpty()) {
                                row.forEach { cell -> header.add(cell.toString()) }
                            } else if (isNotEmptyRow(row)){
                                for (i in 0 until header.size) {
                                    outputRow[header[i]] = cellValue(row.getCell(i))
                                }
                            }
                            if (!outputRow.isEmpty()) {
                                list.add(outputRow)
                            }
                        }
                    }
                }
        writeJson(directory, file, sheets)
        println("Done")
    }
}
